internal/pkg/errorconcept/device_service.go

```diff
@@ -0,0 +1,124 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+package errorconcept
+
+import (
+	"net/http"
+
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+)
+
+var DeviceService deviceServiceErrorConcept
+
+// DeviceServiceErrorConcept represents the accessor for the device-service-specific error concepts
+type deviceServiceErrorConcept struct {
+	EmptyAddressable    deviceServiceEmptyAddressable
+	AddressableNotFound deviceServiceAddressableNotFound
+	NotUnique           deviceServiceNotUnique
+	NotFound            deviceServiceNotFound
+	InvalidState        deviceServiceInvalidState
+}
+
+// NewDeviceServiceDuplicate instantiates a new deviceServiceDuplicate error concept in effort to handle stateful
+// DeviceService duplicate errors
+func NewDeviceServiceDuplicate(currentDSId string, newDSId string) deviceServiceDuplicate {
+	return deviceServiceDuplicate{CurrentDSId: currentDSId, NewDSId: newDSId}
+}
+
+type deviceServiceDuplicate struct {
+	CurrentDSId string
+	NewDSId     string
+}
+
+func (r deviceServiceDuplicate) httpErrorCode() int {
+	return http.StatusConflict
+}
+
+func (r deviceServiceDuplicate) isA(err error) bool {
+	return r.CurrentDSId != r.NewDSId
+}
+
+func (r deviceServiceDuplicate) message(err error) string {
+	return "Duplicate name for Device Service"
+}
+
+type deviceServiceEmptyAddressable struct{}
+
+func (r deviceServiceEmptyAddressable) httpErrorCode() int {
+	return http.StatusBadRequest
+}
+
+func (r deviceServiceEmptyAddressable) isA(err error) bool {
+	panic("isA should not be invoked, this is to only be used as a default error concept")
+}
+
+func (r deviceServiceEmptyAddressable) message(err error) string {
+	return "Must provide an Addressable for Device Service"
+}
+
+type deviceServiceAddressableNotFound struct{}
+
+func (r deviceServiceAddressableNotFound) httpErrorCode() int {
+	return http.StatusNotFound
+}
+
+func (r deviceServiceAddressableNotFound) isA(err error) bool {
+	return err == db.ErrNotFound
+}
+
+func (r deviceServiceAddressableNotFound) message(err error) string {
+	return "Addressable not found by ID or Name"
+}
+
+type deviceServiceNotUnique struct{}
+
+func (r deviceServiceNotUnique) httpErrorCode() int {
+	return http.StatusConflict
+}
+
+func (r deviceServiceNotUnique) isA(err error) bool {
+	return err == db.ErrNotUnique
+}
+
+func (r deviceServiceNotUnique) message(err error) string {
+	return "Duplicate name for the device service"
+}
+
+type deviceServiceNotFound struct{}
+
+func (r deviceServiceNotFound) httpErrorCode() int {
+	return http.StatusNotFound
+}
+
+func (r deviceServiceNotFound) isA(err error) bool {
+	return err == db.ErrNotFound
+}
+
+func (r deviceServiceNotFound) message(err error) string {
+	return "Device service not found"
```

Comment:
Is this the correct message?  Previous implementation was reporting "Addressable not found by ID or Name" at internal/core/metadata/rest_deviceservice.go:70 -- maybe that instance is a different error concept?

Reply:
Good catch! This is a mistake. With my latest, commit to the branch, I have gone through all of the changes with the thought of catching other instances similar to this one. I found 1 or 2 other than this which slipped by and corrected them.